pypesto/visualize/sampling.py

```diff
@@ -91,6 +114,565 @@ def sampling_fval_trace(
     return ax
 
 
+def _get_level_percentiles(level: float) -> Tuple[float, float]:
+    """Convert a credibility level to percentiles.
+
+    Similar to highest-density regions of a normal distribution.
+
+    For example, an credibility level of `95` will be converted to
+    `(2.5, 97.5)`.
+
+    Parameters
+    ----------
+    level:
+        The credibility level used to calculate the percentiles. For example,
+        `[95]` for a 95% credibility interval. These levels are split
+        symmetrically, e.g. `95` corresponds to plotting values between the
+        2.5% and 97.5% percentiles, and are equivalent to highest-density
+        regions for a normal distribution. For skewed distributions, asymmetric
+        percentiles may be preferable, but are not yet implemented.
+
+    Returns
+    -------
+    The percentiles, with the lower percentile first.
+    """
+    lower_percentile = (100 - level) / 2
+    return lower_percentile, 100 - lower_percentile
+
+
+def _get_statistic_data(
+        summary: Dict[str, PredictionResult],
+        statistic: str,
+        condition_index: int,
+        observable_id: str,
+) -> Tuple[Sequence[float], Sequence[float]]:
+    """Get statistic-, condition-, and observable-specific data.
+
+    Parameters
+    ----------
+    summary:
+        A `pypesto.ensemble.EnsemblePrediction.prediction_summary`, used as the
+        source of annotated data to subset.
+    statistic:
+        The predicted statistic, e.g. `MEDIAN` or
+        `get_percentile_label(95)`.
+    condition_index:
+        Select data for a specific condition by its index in a
+        `PredictionResult.condition_ids` object.
+    observable_id:
+        Select data for a specific observable by its ID.
+
+    Returns
+    -------
+    Predicted values and their corresponding time points. A tuple of two
+    sequences, where the first sequence is time points, and the second
+    sequence is predicted values at the corresponding time points.
+    """
+    condition_result = summary[statistic].conditions[condition_index]
+    t = condition_result.timepoints
+    observable_index = condition_result.observable_ids.index(observable_id)
+    y = condition_result.output[:, observable_index]
+    return (t, y)
+
+
+def _plot_trajectories_by_condition(
+        summary: Dict[str, PredictionResult],
+        condition_ids: Sequence[str],
+        observable_ids: Sequence[str],
+        axes: matplotlib.axes.Axes,
+        levels: Sequence[float],
+        levels_opacity: Dict[int, float],
+        labels: Dict[str, str],
+        variables_color: Sequence[RGB],
+) -> None:
```

Comment:
In this and the subsequent function, I'd like to:
 - see more comments on what is actually done and why (what are the quantities which get computed, what is precisely plotted... (what is t0, t1, ...)
 - have some structure, i.e., separation between 1. computing the stuff which is necessary for plotting, 2. plotting things actually, 3. Beautifying by adding axes labels/titles/...
 One could also use subroutines to  group these things together...

Currently, it's understandable somehow, but things could be improved to make it better readable. Don't forget that we have quite some nesting of loops here as well... Subfunction might help.

Reply:
Added comments on the different sections of logic and the unclear variables.